### ISSUE TITLE
Dangerous data format default in SSIM

### ISSUE THREAD

**Issue by khazit** [tensorflow/tensorflow#41650]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Red Hat Linux 7
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.2
- Python version: 3.8
- CUDA/cuDNN version: 10 
- GPU model and memory:

**Describe the current behavior**
[SSIM](https://www.tensorflow.org/api_docs/python/tf/image/ssim) assumes that the image batches have a shape (h, w, batch_size). This is not documented anywhere and the only reason I was aware of it is when I had a batch size smaller than 11, which triggers the following assertion to fail:
``` Python
  checks = [
      control_flow_ops.Assert(
          math_ops.reduce_all(
              math_ops.greater_equal(shape1[-3:-1], filter_size)),
          [shape1, filter_size],
          summarize=8),
      control_flow_ops.Assert(
          math_ops.reduce_all(
              math_ops.greater_equal(shape2[-3:-1], filter_size)),
          [shape2, filter_size],
          summarize=8)
  ]
``` 
Because:
```
>>> shape = (4, 255, 255)
>>> shape[-3:-1]
(4, 255)
```
and `filter_size = 11`

**Describe the expected behavior**
This behaviour should be documented.
The user should have the choice to choose the data format : bath size first or last. By default, I think it should be batch size first as it's the way tf.data.Dataset handles the input pipeline. 

**Standalone code to reproduce the issue**
This works
```
shape = (255, 255, 4)
img1 = tf.zeros(shape)
img2 = tf.zeros(shape)
tf.image.ssim(
    img1, img2, 1
)
```
But this doesn't
```
shape = (4, 255, 255)
img1 = tf.zeros(shape)
img2 = tf.zeros(shape)
tf.image.ssim(
    img1, img2, 1
)
```

**Comment by ravikyram** [comment]
@khazit 

I have tried in colab with TF version 2.2,2.3-rc1.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/b978e4ca53d728e584baa76ce67bfc4b/untitled177.ipynb).You are refering to same behavior?
Thanks!

**Comment by khazit** [comment]
Yes it's the same behavior

**Comment by khazit** [comment]
@jvishnuvardhan any update on this ? I'm willing to make work on it.

**Comment by zhujun98** [comment]
I had the same question these days. After digging into the code 
https://github.com/tensorflow/tensorflow/blob/14246bd0280d1c5e5564eedd4927bf9af9339967/tensorflow/python/ops/image_ops_impl.py#L4307
, it turns out that the implementation assumes the input data has a shape of `(batch size, h, w, channel)`. It calculate the ssim per channel and takes the average for each image, i.e the return is expected to have a shape of `(batch size,)`. 

However, if the input data happens to have a shape of `(batch size, h, w)`. It simply assume the input data has only 1 batch with a shape of `(h, w, channel)` and continue with the calculation. If the `batch_size` happens to be smaller than the `filter_size`, an exception will be raised during the shape check.

I would still call the implementation dangerous since people will get the wrong result with an input data shape of `(batch size, h, w)`, when the `batch_size` is larger than the `filter_size`. It could still work nicely during training as a loss function (1 - ssim).

Finally, the correct way is to reshape the data to `(batch size, h, w, 1)`.

**Comment by jvishnuvardhan** [comment]
@khazit Example in the doc was updated to clearly mention the data shape required for this function. Please check the updated example [here](https://www.tensorflow.org/api_docs/python/tf/image/ssim?version=nightly). Thanks for raising this issue. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/ee5449eed95302b912685d249440cb1d/untitled177.ipynb)

Expected shape of a batch image is (batch_size, height, width, channels). The following works fine.

```
shape = (4, 255, 255,3)
img1 = tf.zeros(shape)
img2 = tf.zeros(shape)
tf.image.ssim(
    img1, img2, 1
)


output 
<tf.Tensor: shape=(4,), dtype=float32, numpy=array([1., 1., 1., 1.], dtype=float32)>
```

I am closing this issue. Please feel free to reopen if I am mistaken. Thanks!
